README.md
# Phoenix Eloquent Meta

[![Latest Version](https://img.shields.io/github/release/chrismichaels84/eloquent-meta.svg?style=flat-square)](https://github.com/phoenix-labs/eloquent-meta/releases)
[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE.md)
[![Build Status](https://img.shields.io/travis/chrismichaels84/eloquent-meta/master.svg?style=flat-square)](https://travis-ci.org/phoenix-labs/eloquent-meta)
[![Total Downloads](https://img.shields.io/packagist/dt/phoenix/eloquent-meta.svg?style=flat-square)](https://packagist.org/packages/phoenix/eloquent-meta)

Attach meta data to [Laravel's](http://laravel.com/) [Eloquent models](http://laravel.com/docs/eloquent).
  * Optionally create a seperate table for each Model
  * Use with or without Laravel
  * Includes Laravel migrations or schema instructions
  * Get meta or fallback

##### Stable Version: 1.4.* works with Laravel 5.* or independently. Pulls Eloquent in automatically.
To use for Laravel 4, see version 1.2.*

## Installation
Install through Composer.

```js
"require": {
    "phoenix/eloquent-meta": "1.4.*"
}
```

That's everything if you are using EloquentMeta and Eloquent **without** using Laravel. You will have to setup Eloquent as detailed in its [documentation](https://github.com/illuminate/database).

If you **are using Laravel**, then you'll want to include the ServiceProvider that will register commands and the like. Update `config/app.php` to include a reference to this package's service provider in the providers array.

```php
'providers' => [
    'Phoenix\EloquentMeta\ServiceProvider'
]
```

Finally, run the migration `php artisan vendor:publish` and `php artisan migrate` to create the database table.

If you **are not using Laravel** then you must create the table manually.

```sql
CREATE TABLE meta
(
    id INTEGER PRIMARY KEY NOT NULL,
    metable_id INTEGER NOT NULL,
    metable_type TEXT NOT NULL,
    key varchar(30) NOT NULL,
    value TEXT NOT NULL,
    locale varchar(30) NULL
);
CREATE UNIQUE INDEX meta_key_index ON meta (key);
CREATE UNIQUE INDEX meta_metable_id_index ON meta (metable_id);

```

## Usage
Add the trait to all models that you want to attach meta data to:

```php
use Illuminate\Database\Eloquent\Model;
use Phoenix\EloquentMeta\MetaTrait;

class SomeModel extends Model
{
    use MetaTrait;

    // model methods
}
```

Then use like this:

```php
$model = SomeModel::find(1);
$model->getAllMeta();
$model->setLocale('th'); // default is 'en'
$model->getMeta('some_key', 'optional default value'); // default value only returned if no meta found.
$model->updateMeta('some_key', 'New Value');
$model->deleteMeta('some_key');
$model->deleteAllMeta();
$model->addMeta('new_key', ['First Value']);
$model->appendMeta('new_key', 'Second Value');
```

### Unique Meta Models and Tables
You can also define a specific meta model for a meta type. For instance, your User model can use UserMeta model with custom methods and all. Using the example above:

```php
use Illuminate\Database\Eloquent\Model;
use Phoenix\EloquentMeta\MetaTrait;

class SomeModel extends Model
{
    use MetaTrait;
    
    protected $meta_model = 'Fully\Namespaced\SomeModelMeta';

    // model methods
}
```
Then in SomeModelMeta simply extends Phoenix\EloquentMeta\Meta. You may now add custom methods to the meta model. You may also dictate which table the metadata is saved to by adding

```php
protected $table = "whatever_table";
```

If you are using EloquentMeta independent of Laravel, then you will have to create the database table manually.

If you are using Laravel, then include the service provider in your config/app.php

```php
'providers' => [
    'Phoenix\EloquentMeta\ServiceProvider'
]
```

Then run ```php artisan generate:metatable table_name``` to create the migration and run ```php artisan migrate``` to build the table.

## Contributing
[Phoenix Labs](http://phoenixlabstech.org) is a non-profit organization developing community-driven experiments in collaboration and content. "Community Driven" is more than just a slogan. It's a core value. Everything we do is for the community and by the community. Collaboration on any of the projects is welcome.

Please se [CONTRIBUTING.md] for more information and for testing.

### Thank you and Credits
Contributors
  - Michael Wilson - @[chrismichaels84](http://github.com/chrismichaels84) - Maintainer
  - Paweł Ciesielski - @[dzafel](http://github.com/dzafel)
  - Lukas Knuth - @[LukasKnuth](http://github.com/LukasKnuth)
 
Many thanks to [Boris Glumpler](https://github.com/shabushabu) and [ScubaClick](https://github.com/ScubaClick) for the original package!



CHANGELOG.md
# Changelog

All Notable changes to `EloquentMeta` will be documented in this file

## NEXT

## v1.4.5 - 4.16.2016
- add locale column to support multi language model

## v1.4.4 - 9.29.2015
- `Helpers::maybeEncode()` does better check of json data. Closes #17

## v1.4.3 - 3.25.2015
- Fixes bug that caused conflict with the given SchemaTemplate and the package migration

## v1.4.2 - 3.25.2015
- Travis setup and CI

## v1.4.1 - 3.24.2015
### Added
- Test Suite
- Best Practice Project Skeleton
- Travis CI
- Contributing Guide and Licence
- Updated Readme

## v1.3.2 
- Make compatible with Laravel 5

tests/EloquentMetaTest.php
<?php
namespace Phoenix\EloquentMeta\Test;

use Illuminate\Database\Capsule\Manager as Capsule;
use Phoenix\EloquentMeta\Test\Stubs\CustomModelParent;
use Phoenix\EloquentMeta\Test\Stubs\Test;
use Phoenix\EloquentMeta\Test\Stubs\TestModelForUniqueMetaModel;

class MetaTest extends \PHPUnit_Framework_TestCase
{
    public $db = false;
    public $testData;

    public function setup()
    {
        if (!$this->db) {
            $this->db = new Capsule;

            $this->db->addConnection([
                'driver'   => 'sqlite',
                'database' => __DIR__ . '/db/test.sqlite',
                'prefix'   => '',
            ]);

            // Make this Capsule instance available globally via static methods... (optional)
            $this->db->setAsGlobal();

            // Setup the Eloquent ORM... (optional; unless you've used setEventDispatcher())
            $this->db->bootEloquent();

            $this->teardown();
            // Tests table
            Capsule::schema()->create('tests', function($table) {
                $table->increments('id');
                $table->string('name')->unique();
                $table->string('email');
            });

            // Meta Table
            Capsule::schema()->create('meta', function($table) {
                $table->increments('id');
                $table->integer('metable_id')->unsigned();
                $table->string('metable_type', 255);
                $table->string('key', 128);
                $table->text('value');
                $table->string('locale', 30)->default('en');

                $table->index('metable_id');
                $table->index('key');
                $table->index('locale');
            });

            Capsule::insert('insert into tests (name, email) values (?, ?)', ['Nicole', 'nicole@nicole.com']);
            Capsule::insert('insert into tests (name, email) values (?, ?)', ['Michael', 'michael@michael.com']);
            Capsule::insert('insert into tests (name, email) values (?, ?)', ['Jon', 'jon@jon.com']);
        }

        $this->testData = new \stdClass();
        $this->testData->nicole = ['id' => 1, 'name' => 'Nicole', 'email' => 'nicole@nicole.com'];
        $this->testData->michael = ['id' => 2, 'name' => 'Michael', 'email' => 'michael@michael.com'];
        $this->testData->jon = ['id' => 3, 'name' => 'Jon', 'email' => 'jon@nicole.com'];
    }

    public function teardown()
    {
        Capsule::schema()->dropIfExists('meta');
        Capsule::schema()->dropIfExists('tests');
    }

    public static function tearDownAfterClass()
    {
        file_put_contents(__DIR__ . "/db/test.sqlite", "");
    }

    public function test_database_setup()
    {
        $result = Capsule::select('select * from tests where name = ?', ['Nicole']);
        $this->assertEquals([$this->testData->nicole], $result);
    }

    public function test_model_setup()
    {
        $result = Test::find(1);
        $this->assertEquals($this->testData->nicole, $result->toArray());
    }

    public function test_add_and_get_meta()
    {
        $model = Test::find(1);
        $model->addMeta('testA', 'testA-value');
        $actual = $model->getMeta('testA');

        $this->assertEquals('testA-value', $actual, "failed to add and get single meta item");
    }

    public function test_get_meta_default_value()
    {
        $model = Test::find(1);
        $actual = $model->getMeta('testB', 'testB-default-value');

        $this->assertEquals('testB-default-value', $actual, "failed to get fallback for single meta item");
    }

    public function test_add_meta_as_array()
    {
        $model = Test::find(1);
        $model->addMeta('testC', ['testC-value', 'testC-second-value']);
        $actual = $model->getMeta('testC');

        $this->assertEquals(['testC-value', 'testC-second-value'], $actual, "failed to get array meta item");
    }

    public function test_update_meta()
    {
        $model = Test::find(1);
        $model->addMeta('testD', 'testD-original-value');
        $model->updateMeta('testD', 'testD-updated-value');
        $actual = $model->getMeta('testD');

        $this->assertEquals('testD-updated-value', $actual, "failed to update meta item");
    }

    public function test_delete_meta()
    {
        $model = Test::find(1);
        $model->addMeta('testE', 'testE-value');
        $model->deleteMeta('testE');
        $actual = $model->getMeta('testD');

        $this->assertNull($actual, "failed to delete meta item");
    }

    public function test_delete_all_meta()
    {
        $model = Test::find(1);
        $model->addMeta('testF-first', 'testF-first-value');
        $model->addMeta('testF-second', 'testF-second-value');
        $model->addMeta('testF-third', 'testF-third-value');

        $model->deleteAllMeta();

        $first = $model->getMeta('testF-first');
        $second = $model->getMeta('testF-second');
        $third = $model->getMeta('testF-third');

        $this->assertNull($first, "failed to delete first meta item");
        $this->assertNull($second, "failed to delete second meta item");
        $this->assertNull($third, "failed to delete third meta item");
    }

    public function test_append_meta()
    {
        $model = Test::find(1);
        $model->addMeta('testG', ['testG-value']);
        $model->appendMeta('testG', 'testG-appended-value');
        $actual = $model->getMeta('testG');

        $this->assertEquals(['testG-value', 'testG-appended-value'], $actual, "failed to get append meta item");
    }

    public function test_get_all_meta()
    {
        $model = Test::find(1);

        $model->addMeta('testH-first', 'testH-first-value');
        $model->addMeta('testH-second', 'testH-second-value');
        $model->addMeta('testH-third', 'testH-third-value');

        $actual = $model->getAllMeta();

        $this->assertEquals([
            'testH-first' => 'testH-first-value',
            'testH-second' => 'testH-second-value',
            'testH-third' => 'testH-third-value',
        ], $actual->toArray(), 'failed to get all meta');
    }

    public function test_model_methods()
    {
        $model = Test::find(1);
        $actual = $model->modelMethod();

        $this->assertEquals('this is a TestModel method', $actual, "failed to get append meta item");
    }

    public function test_custom_meta_model()
    {
        // This test is ridiculously complicated. There has to be an easier way!
        // If tests fail or error, you may have to clear out test.sqlite before running again

        // The `CustomModelParent` model uses a custom model to handle it's meta
        // The `CustomModelChild` handles meta for `CustomModelParent`
        // The `CustomModelChild` uses `custom_model_meta` as its table

        // First, create the `custom_model_parents` table and insert a record
        Capsule::schema()->create('custom_model_parents', function($table) {
            $table->increments('id');
            $table->string('dummy', 255);
        });

        Capsule::insert('insert into custom_model_parents (dummy) values (?)', ['data']);

        // Second, create the `custom_model_meta` table
        Capsule::schema()->create('custom_model_meta', function($table) {
            $table->increments('id');
            $table->integer('metable_id')->unsigned();
            $table->string('metable_type', 255);
            $table->string('key', 128);
            $table->text('value');
            $table->string('locale', 30)->default('en');

            $table->index('metable_id');
            $table->index('key');
            $table->index('locale');
        });

        // Start up the parent
        $model = CustomModelParent::find(1);

        // Now test that the CustomModelParent is working
        $this->assertEquals('this is a CustomModelParent method', $model->modelMethod(), 'failed to use custom method for Parent');

        // Test that data is being saved to the correct table
        $model->addMeta('testTable', 'value');
        $results = Capsule::select('select * from custom_model_meta');

        $this->assertEquals([
            'id' => 1,
            'metable_id' => '1',
            'metable_type' => 'Phoenix\EloquentMeta\Test\Stubs\CustomModelParent',
            'key' => 'testTable',
            'value' => 'value'
        ], $results[0], "failed to get save meta to correct table");

        // Last, clean up our tables
        Capsule::schema()->dropIfExists('custom_model_meta');
        Capsule::schema()->dropIfExists('custom_model_parents');
    }
}


src/Phoenix/EloquentMeta/Meta.php
<?php namespace Phoenix\EloquentMeta;

use Illuminate\Container\Container;
use Illuminate\Support\MessageBag;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Validation\Factory;
use Symfony\Component\Translation\Translator;

class Meta extends Model
{
    /**
     * The database table used by the model.
     *
     * @var string
     */
    protected $table = 'meta';

    /**
     * No timestamps for meta data
     *
     * @var bool
     */
    public $timestamps = false;

    /**
     * Defining fillable attributes on the model
     *
     * @var array
     */
    protected $fillable = [
        'metable_id',
        'metable_type',
        'key',
        'value',
        'locale',
    ];

    /**
     * Error message bag
     *
     * @var \Illuminate\Support\MessageBag
     */
    protected $errors;

    /**
     * Validation rules
     *
     * @var array
     */
    public static $rules = [
        'metable_id'   => 'required|integer',
        'metable_type' => 'required',
        'key'          => 'required|max:100',
        'value'        => 'required',
    ];

   /**
     * Listen for save event
     *
     * @return void
     */
    protected static function boot()
    {
        parent::boot();

        static::saving(function($model) {
            return $model->validate();
        });
    }

    /**
     * Validates current attributes against rules
     *
     * @return bool
     */
    public function validate()
    {
        $validatorFactory = new Factory(new Translator('en'), new Container());
        $validator = $validatorFactory->make($this->attributes, static::$rules);

        if ($validator->passes()) {
            return true;
        }

        $this->setErrors($validator->messages());

        return false;
    }

    /**
     * Set error message bag
     *
     * @var \Illuminate\Support\MessageBag
     * @return void
     */
    protected function setErrors(MessageBag $errors)
    {
        $this->errors = $errors;
    }

    /**
     * Retrieve error message bag
     *
     * @return \Illuminate\Support\MessageBag
     */
    public function getErrors()
    {
        return $this->hasErrors() ? $this->errors : new MessageBag;
    }

    /**
     * Check if a model has been saved
     *
     * @return boolean
     */
    public function isSaved()
    {
        return $this->hasErrors() ? false : true;
    }

    /**
     * Check if there are errors
     *
     * @return boolean
     */
    public function hasErrors()
    {
        return $this->errors instanceof MessageBag;
    }

    /**
     * Connect the models
     * @return \Illuminate\Database\Eloquent\Relations\MorphTo
     */
    public function metable()
    {
        return $this->morphTo();
    }

    /**
     * Maybe decode a meta value
     *
     * @param $value
     * @return mixed
     */
    public function getValueAttribute($value)
    {
        return Helpers::maybeDecode($value);
    }

    /**
     * Maybe encode a value for saving
     *
     * @param $value
     * @return null
     */
    public function setValueAttribute($value)
    {
        $this->attributes['value'] = Helpers::maybeEncode($value);
    }
}


src/Phoenix/EloquentMeta/Helpers.php
<?php namespace Phoenix\EloquentMeta;

class Helpers
{
    /**
     * Checks if a value is json encoded
     *
     * @param $value
     * @param bool $asArray
     * @return mixed
     */
    public static function maybeDecode($value, $asArray = false)
    {
        if (!is_string($value)) {
            return $value;
        }

        $decoded = json_decode($value, $asArray);

        if (json_last_error() == JSON_ERROR_NONE && (is_object($decoded) || is_array($decoded))) {
            return $decoded;
        } else {
            // Reset the JSON error code:
            json_decode("[]");

            return $value;
        }
    }

    /**
     * Checks if a value needs to get json encoded
     *
     * @param $value
     * @return mixed
     */
    public static function maybeEncode($value)
    {
        if (is_array($value) || is_object($value)) {
            $value = json_encode($value, JSON_UNESCAPED_UNICODE);
        }

        return $value;
    }
}


src/Phoenix/EloquentMeta/MetaTrait.php
<?php namespace Phoenix\EloquentMeta;

use Illuminate\Support\Collection;

trait MetaTrait
{
    protected $metaLocale = 'en';

    /**
     * Set meta locale
     *
     * @return \Phoenix\EloquentMeta\MetaTrait
     */
    public function setLocale($locale)
    {
        $this->metaLocale = $locale;
        return $this;
    }

    /**
     * Gets all meta data
     *
     * @return \Illuminate\Support\Collection
     */
    public function getAllMeta()
    {
        return new Collection($this->meta->where('locale', $this->metaLocale)->lists('value', 'key'));
    }

    /**
     * Gets meta data
     *
     * @param $key
     * @param null $default
     * @param bool $getObj
     * @return Collection
     */
    public function getMeta($key, $default = null, $getObj = false)
    {
        $meta = $this->meta()
            ->where('key', $key)
            ->where('locale', $this->metaLocale)
            ->get();

        if ($getObj) {
            $collection = $meta;
        } else {
            $collection = new Collection();

            foreach ($meta as $m) {
                $collection->put($m->id, $m->value, $m->locale);
            }
        }

        // Were there no records? Return NULL if no default provided
        if (0 == $collection->count()) {
            return $default;
        }

        return $collection->count() <= 1 ? $collection->first() : $collection;
    }

    /**
     * Updates meta data
     *
     * @return mixed
     */
    public function updateMeta($key, $newValue, $oldValue = false)
    {
        $meta = $this->getMeta($key, null, true);

        if ($meta == null) {
            return $this->addMeta($key, $newValue);
        }

        $obj = $this->getEditableItem($meta, $oldValue);

        if ($obj !== false) {
            $isSaved = $obj->update([
                'value' => $newValue
            ]);

            return $isSaved ? $obj : $obj->getErrors();
        }

        return null;
    }

    /**
     * Adds meta data
     *
     * @return mixed
     */
    public function addMeta($key, $value)
    {
        $existing = $this->meta()
            ->where('key', $key)
            ->where('locale', $this->metaLocale)
            ->first();

        if ($existing) {
            return false;
        }

        $meta = $this->meta()->create([
            'key'   => $key,
            'value' => $value,
            'locale' => $this->metaLocale,
        ]);

        return $meta->isSaved() ? $meta : $meta->getErrors();
    }

    /**
     * Appends a value to an existing meta entry
     * Resets all keys
     *
     * @return mixed
     */
    public function appendMeta($key, $value)
    {
        $meta = $this->getMeta($key);

        if (!$meta) {
            $meta = [];
        } elseif (!is_array($meta)) {
            $meta = [$meta];
        }

        if (is_array($value)) {
            $meta = array_merge($meta, $value);
        } else {
            $meta[] = $value;
        }

        return $this->updateMeta($key, array_values(array_unique($meta)));
    }

    /**
     * Deletes meta data
     *
     * @param $key
     * @param bool $value
     * @return mixed
     */
    public function deleteMeta($key, $value = false)
    {
        if ($value) {
            $meta = $this->getMeta($key, null, true);

            if ($meta == null) {
                return false;
            }

            $obj = $this->getEditableItem($meta, $value);

            return $obj !== false ? $obj->delete() : false;
        } else {
            return $this->meta()->where('key', $key)->where('locale', $this->metaLocale)->delete();
        }
    }

    /**
     * Deletes all meta data
     *
     * @return mixed
     */
    public function deleteAllMeta()
    {
        return $this->meta()->delete();
    }

    /**
     * Gets an item to edit
     *
     * @return mixed
     */
    protected function getEditableItem($meta, $value)
    {
        if ($meta instanceof Collection) {
            if ($value === false) {
                return false;
            }

            $filtered = $meta->filter(function ($m) use ($value) {
                return $m->value == $value && $m->locale == $this->metaLocale;
            });

            $obj = $filtered->first();

            if ($obj == null) {
                return false;
            }
        } else {
            $obj = $meta;
        }

        return $obj->exists ? $obj : false;
    }

    /**
     * Attaches meta data
     *
     * @return object
     */
    public function meta()
    {
        $meta_model = isset($this->meta_model) ? $this->meta_model : 'Phoenix\EloquentMeta\Meta';
        return $this->morphMany($meta_model, 'metable');
    }
}


src/Phoenix/EloquentMeta/SchemaTemplate.php
<?php

use Illuminate\Database\Schema\Blueprint;
use Illuminate\Database\Migrations\Migration;

class SchemaTemplate extends Migration
{
    /**
     * Run the migrations.
     *
     * @return void
     */
    public function up()
    {
        Schema::create('__meta__', function(Blueprint $table) {
            $table->increments('id');
            $table->integer('metable_id')->unsigned();
            $table->string('metable_type', 255);
            $table->string('key', 128);
            $table->text('value');
            $table->string('locale', 30)->default('en');

            $table->index('metable_id');
            $table->index('key');
            $table->index('locale');
        });
    }

    /**
     * Reverse the migrations.
     *
     * @return void
     */
    public function down()
    {
        Schema::drop('__meta__');
    }
}
